File: Coding Interviews/16_ReverseList/ReverseList.py
'''
Copyright: Amusi
Author:    Amusi
Date:      2018-06-26

题目描述
输入一个链表，反转链表后，输出新链表的表头。


'''

# -*- coding:utf-8 -*-
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    # 返回ListNode
    # 思路: 即将链表进行反转
    def ReverseList(self, pHead):
        # write code here
        # 输入判断
        if pHead == None or pHead.next == None:
            return pHead
        pre = None
        cur = pHead
        while cur:
            post = cur.next
            cur.next = pre
            pre = cur
            cur = post
        return pre

File: Coding Interviews/16_ReverseList/ReverseList.cpp
/*****************************************
Copyright: Amusi
Author:    Amusi
Date:      2018-06-26

题目描述
输入一个链表，反转链表后，输出新链表的表头。

*****************************************/

/*
struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};*/
class Solution {
public:
    // 思路: 将列表进行反转
    ListNode* ReverseList(ListNode* pHead) {
        if(pHead==NULL || pHead->next==NULL)
            return pHead;
        
        ListNode* cur = pHead;
        ListNode* pre = NULL;    // ListNode* pre = NULL;
        ListNode* post = NULL;    // ListNode* post = pHead->next;
        while (cur!=NULL){
            post = cur->next;
            cur->next = pre;
            pre = cur;
            cur = post;
        }
        pHead = pre;
        return pHead;
    }
};